proposals/p2240.md

```diff
@@ -588,85 +604,198 @@ rejected), so we will focus on the semantics of phase 3 when the pattern type is
+there are N non-variadic leading argument expressions. Likewise, there will
+always be at least M trailing arguments. As a result, a leading parameter can
+only match a leading argument, and so it can only match a leading argument
+expression, and likewise for trailing parameters. Consequently, a leading
+argument expression cannot match a trailing parameter, a trailing argument
+expression cannot match a leading parameter, and a central argument expression
+can only match the variadic parameter.
+
+Consider the i'th non-variadic leading argument expression E. If all the
+variadic argument expressions before it are empty, E will match the i'th leading
+parameter, so E cannot match any earlier parameter. If there are any earlier
+variadic argument expressions, E can be made to match any later leading
+parameter or the variadic parameter, by making one of those earlier variadic
+arguments large enough, but as observed above, E cannot match a trailing
+parameter. If there are no earlier variadic argument expressions, E cannot be
+made to match any later parameter, so it can only match the i'th leading
+parameter.
+
+Next, consider a variadic leading argument expression E that comes before the
+i'th non-variadic leading argument expression, but not before any earlier
+non-variadic argument expression. If E's rewritten size is sufficiently large,
+and all earlier variadic argument expressions are empty, it will simultaneously
+match the i'th leading parameter, all leading parameters after it, and the
+variadic parameter, but as before, it cannot match a trailing parameter.
+
+The same reasoning can be applied to trailing argument expressions, but with
+"before" and "after" reversed. And as noted earlier, central argument
+expressions can only match the variadic parameter. In summary, we can identify
+the possible matches for an argument expression E as follows:
+
+-   If E is leading, let i be one more than the number of earlier non-variadic
+    argument expressions:
+    -   If E is non-variadic, and there are no earlier variadic argument
+        expressions, then E can only match the i'th leading parameter.
+    -   Otherwise, E can match the i'th leading parameter, any later leading
+        parameters, and the variadic parameter.
+-   If E is trailing, let i be one more than the number of later non-variadic
+    argument expressions:
+    -   If E is non-variadic, and there are no later variadic argument
+        expressions, then E can only match the i'th trailing parameter from the
+        end.
+    -   Otherwise, E can match the i'th trailing parameter from the end, any
+        earlier trailing parameters, and the variadic parameter.
+-   Otherwise, E can only match the variadic parameter.
+
+###### The type-checking algorithm
+
+In order to avoid type deduction that splits into multiple cases, we require
+that if the variadic parameter has a deduced type that is used in more than one
```

Comment:
Isn't this only when it a type pack? Or is the type only deduced when it is a pack?

Reply:
Yeah, I don't think there's any way to deduce a non-pack type for the variadic parameter. In principle I think it could have a type like `[T;]` where `T` is deduced, but I wouldn't call a type with a deduced component a "deduced type". I'm open to suggestions for how to make that clearer, though. In any event, for now we shouldn't allow that, since we don't allow its non-variadic counterpart (I've added a note about that up above).

Reply:
I think it would be clearer if the text said the word "pack" to clarify that is what is being discussed. I don't think `T` can be deduced from a `[T;]` type expression, since it could have length 0. In the `Min` case, the `first` parameter has type `T` and can be used to ensure that `T` can be deduced from the arguments. It is a requirement that every deduced parameter is guaranteed to be deducible from the arguments (and not the return type). I think that means a parameter with a `[T;]` type must be paired with some other parameter that can be used to deduce `T`.

Reply:
I've tweaked this rule to be a bit more general, which I think makes this moot.